src/osp/hierarchical_bitset.h

```diff
@@ -0,0 +1,607 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2021 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include <array>
+#include <bitset>
+#include <cstdint>
+#include <memory>
+#include <stdexcept>
+
+#if defined(_MSC_VER)
+    #include <intrin.h>
+#endif
+
+namespace osp
+{
+
+#ifdef __GNUC__
+
+    inline int ctz(uint64_t a) noexcept { return __builtin_ctzll(a); }
+
+#elif defined(_MSC_VER)
+
+    inline int ctz(uint64_t a) noexcept
+    {
+        unsigned long int b = 0;
+        _BitScanForward64(&b, a);
+        return b;
+    }
+
+#elif
+
+    static_assert(false, "Missing ctz 'count trailing zeros' implementation for this compiler");
+
+#endif
+
+/**
+ * @brief Divide two integers and round up
+ */
+template<typename NUM_T, typename DENOM_T>
+constexpr decltype(auto) div_ceil(NUM_T num, DENOM_T denom) noexcept
+{
+    static_assert(std::is_integral_v<NUM_T>);
+    static_assert(std::is_integral_v<DENOM_T>);
+
+    return (num / denom) + (num % denom != 0);
+}
+
+template<typename INT_T>
+constexpr bool bit_test(INT_T block, int bit)
+{
+    static_assert(std::is_unsigned_v<INT_T>);
+    return (block & (INT_T(1) << bit) ) != 0;
+}
+
+/**
+ * @brief Copy a certain number of bits from one int array to another
+ *
+ * This function treats integer arrays as bit arrays. Bits are indexed from
+ * LSB to MSB.
+ *
+ * TODO: maybe add a starting bit offset too?
```

Comment:
This is definitely an endianness hazard, since the first N bits may not correspond to the least significant bits on all machines. Is `INT_T` meant to mostly be `uint8_t` (i.e. basically a stand-in for a raw byte array), or is this intended to be used with larger integer types?

Reply:
There are no operations that access the individual bytes of the integer. This function will work as long as left-shifting moves to the next bit. A 4 byte integer will always be accessed as 4 bytes. There's no point in using 8-bits ints, since the registers on most machines we're targeting are much larger than 8 bits.

Reply:
Though it would be cool if you find a way to break this in a unit test or something.

Reply:
Alright makes sense, in that case I suppose it will be alright. I guess I'm just thinking too closely about the underlying memory layout. Endianness will still be an issue for any serialization that happens but then again, we'll have plenty of other issues on that front anyways.